cirq/study/flatten_expressions.py

```diff
@@ -0,0 +1,368 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Resolves symbolic expressions to unique symbols."""
+
+from typing import (overload, Any, Callable, Dict, Iterator, List, Optional,
+                    Tuple, Union)
+
+import sympy
+
+from cirq import protocols, value
+from cirq.study import resolver, sweeps, sweepable
+
+
+def flatten(val: Any) -> Tuple[Any, 'ExpressionMap']:
+    """Creates a copy of `val` with any symbols or expressions replaced with
+    new symbols.  `val` can be a `Circuit`, `Gate`, `Operation`, or other
+    type.
+
+    Use `ExpressionMap.transform_sweep` or `ExpressionMap.transform_params`
+    after `flatten`.  Use `flatten_with_sweep` or `flatten_with_params` instead
+    of `flatten` as a shortcut.
+
+    Args:
+        val: The value to copy with substituted parameters.
+
+    Example:
```

Comment:
It seems to have the args and return value separated by the example, is this something being enforced by our doc test?

Reply:
No.  The doc test just looks for '>>>' and following lines.  The example should come before Args though.

Reply:
Fixed.